### PR TITLE
Takeoff Revision

### PR DESCRIPTION
**Pull request by MaEtUgR**
**Describe problem solved by the proposed pull request**
Addressed Problems:
1. The ramp logic was completely broken, see #12001
1. The ramp didn't start at the correct value depending on vertical velocity control tuning
1. The ramp didn't stop at the correct value when providing velocity and position setpoints during takeoff (which smooth mission does)
1. The takeoff could be triggered accidentally when the state fluctuates during idling
1. There was a race condition when starting a takeoff while being disarmed in manual mode but with the throttle stick not at the minimum. In this case the takeoff was completely skiped and it jumped up very agressively.
1. The whole logic was scattered over the position control module and very hard to debug

**Describe your preferred solution**
1. I restored the original logic in #12001 but faced all the other listed problems and therefore continued in this pr.
1. The ramp now starts at the precalculated verticalvelocity setpoint that will result in a thrust setpoint starting at zero value. see `generateInitialValue()` in the code.
1. The ramp now stops at the velocity limit given by the constraint that will anyways be used during flight and is applied to the sum of the position controller output and velocity setpoint feed forward. see `_takeoff.updateRamp(_dt, constraints.speed_up)` where the constraint is now the goal value of the ramp.
1. The initial idea of the flight task architecture was that a task can freely set it's setpoints and doesn't have to worry about takeoff and landing. It would just takeoff when it's landed and there's a setpoint to go up and land when it puts a setpoint that pushes into the ground. With the takeoff logic there are some significant interface problems depending on the way a task is implemented: From the setpoint is not high enough to trigger the takeoff to an unexpected takeoff because of some estimator fluctuation is affecting the setpoint.
It's solved by allowing the task to determine when a takeoff is triggered. If no condition is implemented by default a task is not allowing a takeoff  and cannot be used to get the vehicle off the ground.
See the `constraints.want_takeoff` flag determined by the task to allow a takeoff, it's ignored in flight.
1. As a hotfix I cut the throttle value when disarmed in manual mode. This is an in between solution and we should be able to remove it again when we allow a task to run when disarmed such that it already prepares the right setpoints.
1. To even make it workable solving all these problems I created a Takeoff class which runs contains a state machine originally designed by @RomanBapst and @bresch and the ramp. This makes it a lot clearer what is takeoff logic and not other in flight logic. Also the state machine makes debugging very easy because when you look at the state and the single condition that blocks/skips you know exactly what goes wrong. It also prevents certain corner cases. Probably even 5., that has to be tested.

**Test data / coverage**
I SITL tested and real world tested all the changes in between and that's also how I found all the unexpected additional problems. The last commit is not real world tested yet so that needs to be carefully done. I'll update when I arrived to test it.

**Additional context**
old pr to fix for smooth takeoff: #12001

### CODE REVIEW

**Comment by MaEtUgR** [comment]
CI was very useful and found some initialization and offboard issues. I fixed them.

File: src/lib/FlightTasks/tasks/ManualAltitude/FlightTaskManualAltitude.cpp [hunk 1]
```diff
@@ -345,10 +345,17 @@ void FlightTaskManualAltitude::_updateSetpoints()
 	_respectGroundSlowdown();
 }
 
+bool FlightTaskManualAltitude::_checkTakeoff()
+{
+	// stick is is deflected up
+	return _sticks(2) > 0.65f;
```
Comment (Stifael) [inline src/lib/FlightTasks/tasks/ManualAltitude/FlightTaskManualAltitude.cpp]: in this case the flag `want_to_takeoff` is always true?

**Comment by Stifael** [comment]
Since the takeoff logic is anyway moved into flighttask, another alternative is to let flighttask know about landing state. Then the takeoff-logic does no longer be split between flighttask- and mc-pos-control-main file.

File: src/lib/FlightTasks/tasks/ManualAltitude/FlightTaskManualAltitude.cpp (hunk 1 above)
Comment (MaEtUgR) [reply src/lib/FlightTasks/tasks/ManualAltitude/FlightTaskManualAltitude.cpp]: Yes, but it gets ignored when you're already flying.

**Comment by MaEtUgR** [comment]
@Stifael Thanks a lot for having a look and the valuable input! Note that the thrust ramp I'm currently testing now has the correct coordinate frame. It's a good idea to inform the task of the takeoff state since we can then probably get rid of the reactive problems.

I'm continuing to further improve the takeoff on this branch since https://github.com/PX4/Firmware/pull/12001 is the minimum viable solution release blocker.

**Comment by MaEtUgR** [comment]
I'm working on a takeoff class that has a state machine for the vehicle to pass through (credit to @RomanBapst for that part!) and does a nice thrust ramp. Here's an example takeoff landing with the jerk limited trajectory auto implementation (default):
Takeoff class
Commiting the first version soon.

**Comment by MaEtUgR** [comment]
Real world testing revealed a problem. Please hold back of any real-world testing until I fixed them.

**Comment by MaEtUgR** [comment]
I went outside testing again with a better behaving vehicle and it worked well except for a really small sideways twitch at the end of the ramp because the sideways corrections kick in. I was in discussion with @RomanBapst how to also eliminate that.

Here are logs:
Altitude mode: https://logs.px4.io/plot_app?log=67df14e0-ea90-4e83-9540-9ff56c50d49c
Position mode: https://logs.px4.io/plot_app?log=a1609347-417c-48ba-b10d-19ef98f34e87
Auto takeoff and land: https://logs.px4.io/plot_app?log=e266f409-7897-4003-9b43-fa93e80ced26
https://logs.px4.io/plot_app?log=6509454c-d40f-4b60-a081-ebaf52a9cd10
Auto takeoff position land: https://logs.px4.io/plot_app?log=04b930ff-f6d3-4fe6-b961-b970e146b5fe

I just rebased in the hope that fixes the CI does not build anything problems. I'm uploading videos of the flight as I'm writing this and will post links by editing this comment.

**Comment by LorenzMeier** [comment]
Awesome, thanks!

**Comment by MaEtUgR** [comment]
I wasn't entirely pleased with the pure thrust ramp because for some part of the ramp up time the vehicle ramps nicely and isn't moving but at some point the drone accelerates up quite quickly. Ramping the thrust while the vehicle is still on ground makes sense since it's not yet moving up anyways but as soon as the vehicle left the ground ramping further to reach the current thrust setpoint results in linearly ramping the acting vertical acceleration and hence the vertical velocity goes up quadratic. This rapid increase in velocity is audible and visible and looks like an aggressive maneuver just after leaving the ground. The velocity ramp results in a thrust ramp while the velocity is still on the ground but as soon as the vehicle is moving and the estimated vertical velocity changes the controller takes care of not shooting up but rather go up with constant acceleration until the takeoff speed is met.

Early this morning I had good ideas how to resolve the issues we faced with the velocity ramp and implemented them in the last commit.
- When to begin the ramp?
There's a calculation now for the velocity ramp initial value such that the resulting thrust is zero at the beginning. The result is a fixed real-world takeoff time independent of the vehicle tuning assuming it's able to at least loosely track a velocity setpoint with it's tuning.
- When to end the ramp?
The ramp is applied to the upwards velocity constraint and it just ramps from the initial value to the velocity constraint which is applied during flight. Slower/going down is always possible. The problem we faced with the AutoVelSmooth task which generates a position setpoint and a velocity feed forward was that we used only the feed forward to determine when the ramp is finished. But since the vertical velocity constraint gets applied to the sum inside the position controller and there's always a maximum velocity constraint we can just ramp the constraint up linearly until it's at the value it anyways has during flight.

@bresch @RomanBapst Since I don't have a vehicle at hand **does aynone dare to test this on the real vehicle?** I tested in SITL and it all looks good. I also checked that the only horizontal thrust leading to 90 desired tilt which screwed me yesterday is not happening anywhere.

**Comment by RomanBapst** [comment]
Will test!

**Comment by MaEtUgR** [comment]
I updated the description such that we are all on the same page, sorry for not doing that earlier. The test of the last commit is still pending.

**Comment by RomanBapst** [comment]
@MaEtUgR I will get to test this either today or tomorrow.

**Comment by LorenzMeier** [comment]
@RomanBapst Keep us posted! The last item blocking the release!

**Comment by MaEtUgR** [comment]
I'm looking for a hole in the rain to get this tested for the entire day already but there's no chance 🌧

**Comment by jorge789** [comment]
### Indoor flight test on Pixhawk 2 Cube V3

Modes Tested

Takeoff: Good
Stabilized Mode: Good.

Pre-flight Procedure:

Flight Test: 
Arm and Take off in the stabilized mode ✓
Throttle Test ✓
Pitch, Roll, Yaw ✓
Land ✓

### Notes: Armed and took off in stabilized mode, flew for 30 seconds approximately then proceeded to land and disarm. This process was repeated one after another to a total of 5 take offs.  

Flight log:
https://review.px4.io/plot_app?log=51dd999e-b62b-49df-955e-c4db61ff2e3e


### Indoor flight test on PixRacer V4

Modes Tested

Takeoff: Good
Stabilized Mode: Good.

Pre-flight Procedure:

Flight Test: 
Arm and Take off in the stabilized mode ✓
Throttle Test ✓
Pitch, Roll, Yaw ✓
Land ✓

### Notes: Armed and took off in stabilized mode, flew for 30 seconds approximately then proceeded to land and disarm. This process was repeated one after another to a total of 5 take offs.  

Flight log: 1
https://review.px4.io/plot_app?log=5318fe9f-d710-46b5-b6d5-4a6119d41d81

Flight log: 2
https://review.px4.io/plot_app?log=4c82cb60-a950-47c3-a993-f6f54d2355da

**Comment by dannyfpv** [comment]
tested on pixhawk 4 v5
indoor test
Modes Tested

Takeoff: Good
Stabilized Mode: Good.

Pre-flight Procedure:

Flight Test:
Arm and Take off in the stabilized mode no issue
Throttle Test no issue
Pitch, Roll, Yaw no issue
Land no issue

Notes: Armed and took off in stabilized mode, flew for 30 seconds approximately then proceeded to land and disarm. This process was repeated one after another to a total of 5 take offs.

https://review.px4.io/plot_app?log=2a63b1ed-0800-4b97-8913-f943b37c6793

https://review.px4.io/plot_app?log=025c0d46-cbd5-4fe4-b735-1d7aa25cd17d

**Comment by MaEtUgR** [comment]
When it was not raining for a moment @RomanBapst and I went out and did a couple of tests:
https://drive.google.com/file/d/12IFqDlUiIoZwSqX32CW4kz_79VtkRYFN

Generally the takeoff worked really well. We found two remaining smaller issues:
- When the controller wants to correct sideways the vehicle does that as soon as possible directly when lifting off. This was intentionally to not have any drift at the beginning or discontiuous change in air but could depending on the use case be seen as an issue.
- https://github.com/PX4/Firmware/pull/12014/commits/0c81a19decde6ddfe4ce87c34c762ea15fd3ab09 only fixes the high thrust at the beginning problem described here: https://github.com/PX4/Firmware/pull/12001#issuecomment-491539137 when starting in manual mode. There are very specific corner cases in which you might disarm with higher than idle thrust and then the thrust setpoint stays like this because no task is running and the setpoint does not get published until the pilot arms again. The first setpoint sample then still has the value from before it last disarmed. But it cannot as a result accidentally take off anymore because of the consistent takeoff state machine.

File: src/modules/mc_pos_control/Takeoff/Takeoff.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		_takeoff_state = TakeoffState::flight;
+	}
+
+	// TODO: need to consider free fall here
+	if (!armed) {
+		_takeoff_state = TakeoffState::disarmed;
+	}
+}
+
+float Takeoff::updateRamp(const float dt, const float takeoff_desired_vz)
```
Comment (RomanBapst) [inline src/modules/mc_pos_control/Takeoff/Takeoff.cpp]: @MaEtUgR Let's add a description to this function:
- this function is responsible for ramping up a reference vertical velocity setpoint (_takeoff_ramp_vz) during takeoff
- the reference velocity at the beginning of a takeoff is initialized to a value which when passed to the velocity controller results in a zero throttle setpoint
- by ramping up _takeoff_ramp_vz during the takeoff and using it to constain the maximum climb rate demanded by the active flight task a smooth takeoff behavior is achieved

File: src/modules/mc_pos_control/mc_pos_control_params.c [hunk 3]
```diff
@@ -668,10 +668,10 @@ PARAM_DEFINE_FLOAT(MPC_LAND_ALT2, 5.0f);
  * A time constant of 0 disables the ramp
  *
  * @min 0
- * @max 1
+ * @max 5
  * @group Multicopter Position Control
  */
-PARAM_DEFINE_FLOAT(MPC_TKO_RAMP_T, 0.4f);
+PARAM_DEFINE_FLOAT(MPC_TKO_RAMP_T, 3.0f);
```
Comment (RomanBapst) [inline src/modules/mc_pos_control/mc_pos_control_params.c]: @MaEtUgR Do you think 3 seconds is not a bit too much?

Comment (MaEtUgR) [reply src/modules/mc_pos_control/mc_pos_control_params.c]: I found it to be a good general default value. It's more on the slower side but still nice for a racer and probably desired on any bigger vehicle. What were your impressions? These were the values we tested outside yesterday and reflected in the video from https://github.com/PX4/Firmware/pull/12014#issuecomment-494360978

File: src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp [hunk 4]
```diff
@@ -57,10 +57,12 @@ bool FlightTaskAutoLineSmoothVel::activate()
...
-	// Don't reset during takeoff TODO: Find a proper solution
-	// The issue here is that with a small increment of velocity setpoint (generated by this flight task), the
-	// land detector doesn't detect takeoff and without takeoff detection, the
-	// flight task is always reset.
+	// On ground, reset acceleration and velocity to zero
+	for (int i = 0; i < 2; ++i) {
+		_trajectory[i].reset(0.f, 0.f, _position(i));
+	}
+
+	_trajectory[2].reset(0.f, 0.7f, _position(2));
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp]: It was added to work nicer with the takeoff logic but should actually not matter anymore we can do tests without it and see if it matters but I would suggest to just leave it for now.

File: src/modules/mc_pos_control/Takeoff/Takeoff.cpp (hunk 2 above)
Comment (MaEtUgR) [reply src/modules/mc_pos_control/Takeoff/Takeoff.cpp]: I amended it a bit and also made a description for `generateInitialValue()` because that's where part of the description belongs.

File: src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.hpp [hunk 5]
```diff
@@ -67,6 +67,9 @@ class FlightTaskAutoLineSmoothVel : public FlightTaskAutoMapper2
 
 	void _generateSetpoints() override; /**< Generate setpoints along line. */
 
+	/** determines when to trigger a takeoff (ignored in flight) */
+	bool _checkTakeoff() override { return _want_takeoff; };
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.hpp]: The function does what the name and comment says. In this particular case it looks like a getter but it is the interface to actually decide and where your conditions might go.

**Comment by MaEtUgR** [comment]
https://github.com/PX4/Firmware/pull/12014/commits/fce056a5db16a1774b70314353032ec8651f9f07 fixes the second point of https://github.com/PX4/Firmware/pull/12014#issuecomment-494360978 and everything described in https://github.com/PX4/Firmware/pull/12001#issuecomment-491539137. I reverted https://github.com/PX4/Firmware/commit/0c81a19decde6ddfe4ce87c34c762ea15fd3ab09 because it's not necessary with the new more general fix for more cases.

We went out testing again and did find any way to enter this state anymore. We also tried to break the logic with "landing halfway" and so on but it was all expected behavior. When you go full sideways there is a small twitch when the ramp finishes because before the tilt is constrained.
Here are some logs:
https://logs.px4.io/plot_app?log=0f557e1e-e392-42ad-84a8-eb28b0aa03ad
https://logs.px4.io/plot_app?log=751a2dab-067a-440b-9460-21cedaeaaa7d
https://logs.px4.io/plot_app?log=f9646e05-65fe-46d7-be30-8eb7c5cce73b
https://logs.px4.io/plot_app?log=257769d3-f205-4d64-92f9-cebb6bdff20b
https://logs.px4.io/plot_app?log=69e2edbf-685c-400b-b8b7-831b8486fab6
https://logs.px4.io/plot_app?log=fc083da3-c8e5-42ef-a8b9-e74ca37f85ee
https://logs.px4.io/plot_app?log=ff4c9924-ca4a-4cb3-bd0a-7d24ed82b866
https://logs.px4.io/plot_app?log=926f174b-8bc3-4dcc-8cfd-5a6d1a7265b1
https://logs.px4.io/plot_app?log=a0d31e17-647d-43c3-a447-6e5cfc834a1c
https://logs.px4.io/plot_app?log=7ab7c1e9-bba6-42d6-a1f1-fdf01c3b020a
https://logs.px4.io/plot_app?log=f7d94759-933c-42bf-948a-d19261d30764
https://logs.px4.io/plot_app?log=40b54a0d-8059-420d-b9ac-320f77b53ba7
https://logs.px4.io/plot_app?log=18f972c1-8f5d-4a2a-9462-000e5d065739
https://logs.px4.io/plot_app?log=0c63876e-8275-4f42-992b-27048c47dcd1
https://logs.px4.io/plot_app?log=6712d240-ee56-4a04-a378-f5845301c431
https://logs.px4.io/plot_app?log=4f0f7800-abe1-4f6c-994a-206d872de8b3
https://logs.px4.io/plot_app?log=80c3f123-40cc-4224-a7b8-d6df4378575d
We did a total of > 60 takeoffs with two vehicles in different modes and some smaller flight tests in between.
